cloudsmith/data_source_package.go

```diff
@@ -44,12 +67,77 @@ func dataSourcePackageRead(d *schema.ResourceData, m interface{}) error {
 	d.SetId(fmt.Sprintf("%s_%s_%s", namespace, repository, pkg.GetSlugPerm()))
 
 	if download {
-		outputPath, err := downloadPackage(pkg.GetCdnUrl(), downloadDir, pc)
+		outputPath, err := downloadPackage(pkg.GetCdnUrl(), downloadDir, pc, false)
 		if err != nil {
 			return err
 		}
 		d.Set("output_path", outputPath)
 		d.Set("output_directory", downloadDir)
+
+		// Calculate checksums for the downloaded file
+		localChecksums, err := calculateChecksums(outputPath)
+		if err != nil {
+			return err
+		}
+
+		localMD5 := localChecksums.MD5
+		localSHA1 := localChecksums.SHA1
+		localSHA256 := localChecksums.SHA256
+		localSHA512 := localChecksums.SHA512
+
+		// Check against API checksums
+		if localMD5 != pkg.GetChecksumMd5() || localSHA1 != pkg.GetChecksumSha1() || localSHA256 != pkg.GetChecksumSha256() || localSHA512 != pkg.GetChecksumSha512() {
+			// Checksum doesn't match, try to download again with bustCache set to true
+			outputPath, err := downloadPackage(pkg.GetCdnUrl(), downloadDir, pc, true)
+			if err != nil {
+				return err
+			}
+			fmt.Println("Package pulled again with bustCache due to checksum mismatch.")
+
+			// Calculate checksums for the downloaded file again
+			localChecksums, err := calculateChecksums(outputPath)
```

Comment:
suggestion: The `calculateChecksums` function can be part of the `Checksums` struct and internally modify the values as to avoid re-assigning each time here.